rllib/algorithms/ppo/utils/env_runner.py

```diff
@@ -0,0 +1,471 @@
+import gymnasium as gym
+import numpy as np
+import tree
+
+from functools import partial
+from typing import Dict, List, Optional, Tuple
+from collections import defaultdict
+from ray.experimental.tqdm_ray import tqdm
+from ray.rllib.algorithms.algorithm_config import AlgorithmConfig
+from ray.rllib.core.models.base import STATE_IN, STATE_OUT
+from ray.rllib.core.rl_module.marl_module import MultiAgentRLModule
+from ray.rllib.evaluation.metrics import RolloutMetrics
+from ray.rllib.env.env_runner import EnvRunner
+from ray.rllib.env.utils import _gym_env_creator
+from ray.rllib.policy.sample_batch import DEFAULT_POLICY_ID, SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.framework import try_import_tf
+from ray.rllib.utils.replay_buffers.episode_replay_buffer import _Episode as Episode
+from ray.rllib.utils.rl_module import foreach_module
+from ray.tune.registry import ENV_CREATOR, _global_registry
+
+_, tf, _ = try_import_tf()
+
+
+class PPOEnvRunner(EnvRunner):
+    """An environment runner to collect data from vectorized gymnasium environments."""
+
+    def __init__(self, config: AlgorithmConfig, **kwargs):
+        """Initializes a PPOEnvRunner.
+
+        Args:
+            config: The config to use for setup of this EnvRunner.
+        """
+
+        super().__init__(config=config)
+
+        self.worker_index = kwargs.get("worker_index")
+        # Register env for the local context here.
+        gym.register(
+            "ppo-custom-env-v0",
+            partial(
+                _global_registry.get(ENV_CREATOR, self.config.env),
+                self.config.env_config,
+            )
+            if _global_registry.contains(ENV_CREATOR, self.config.env)
+            else partial(
+                _gym_env_creator,
+                env_context=self.config.env_config,
+                env_descriptor=self.config.env,
+            ),
+        )
+        # Create the vectorized gymnasium env.
+        self.env: gym.vector.VectorEnv = gym.vector.make(
+            "ppo-custom-env-v0",
+            num_envs=self.config.num_envs_per_worker,
+            asynchronous=False,  # self.config_remote_envs
+        )
+        self.num_envs: int = self.env.num_envs
+        assert self.num_envs == self.config.num_envs_per_worker
+
+        # Create the PPORLModule to compute actions with.
+        if self.config.share_module_between_env_runner_and_learner:
+            self.module = None
+        # Create our won instance of a PPORLModule (which then needs
+        # to be weight-synched each iteration).
+        else:
+            policy_dict, _ = self.config.get_multi_agent_setup(env=self.env)
+            module_spec = self.config.get_marl_module_spec(policy_dict=policy_dict)
+            # TODO (simon): This here is only for single agent.
+            # This is a MARL.
+            self.module: MultiAgentRLModule = module_spec.build()  # [DEFAULT_POLICY_ID]
```

Comment:
Maybe we should rename the property, then into `self.marl_module` to have 100% clarity for anything using this EnvRunner (e.g. the Algorithm).